packages/api-request-builder/src/default-params.js

```diff
@@ -62,23 +63,104 @@ export function getDefaultSearchParams (): ServiceBuilderDefaultParams {
 export function setDefaultParams () {
   this.params.expand = getDefaultQueryParams().expand
 
-  if (this.features.indexOf(features.queryOne) >= 0)
+  if (this.features.includes(features.queryOne))
     this.params.id = getDefaultQueryParams().id
 
-  if (this.features.indexOf(features.query) >= 0) {
+  if (this.features.includes(features.query)) {
     this.params.pagination = getDefaultQueryParams().pagination
     this.params.query = getDefaultQueryParams().query
   }
 
-  if (this.features.indexOf(features.search) >= 0) {
+  if (this.features.includes(features.search)) {
     this.params.staged = getDefaultSearchParams().staged
     this.params.pagination = getDefaultSearchParams().pagination
     this.params.search = getDefaultSearchParams().search
   }
 
-  if (this.features.indexOf(features.projection) >= 0)
+  if (this.features.includes(features.projection))
     this.params.staged = true
 
-  if (this.features.indexOf(features.suggest) >= 0)
+  if (this.features.includes(features.suggest))
     this.params.searchKeywords = []
 }
+
+const hasKey = (obj: {}, key: string): boolean =>
+  Object.prototype.hasOwnProperty.call(obj, key)
+
+/**
+ * Set the supplied parameters given the current service object.
+ *
+ * @return {void}
+ */
+export function setParams (params: ServiceBuilderParams) {
+  // verify params
+  const knownKeys = [
+    'expand', 'id', 'key', 'customerId', 'sort', 'page', 'perPage', 'staged',
+    'priceCurrency', 'priceCountry', 'priceCustomerGroup', 'priceChannel',
+    'text', 'fuzzy', 'fuzzyLevel', 'markMatchingVariants', 'facet', 'filter',
+    'filterByQuery', 'filterByFacets', 'searchKeywords', 'where', 'version',
+  ]
+  Object.keys(params).forEach((key: string) => {
+    if (!knownKeys.includes(key)) throw new Error(`Unknown key "${key}"`)
+  })
```

Comment:
Not sure whether we want to warn on unknown keys or not. But it's in for now as it can catch possible errors early. Otherwise we'd fail silently.

Please advise on which is better here. As far as I could tell we're failing loudly in other cases, e.g. when page is provided with `0` (it must be at least `1`).

Reply:
I think in the case of

- where a query is invalid => throw an error
    - e.g `id`(customer-id, product-id) are strings in the API, but the dev can potentially pass in a number. This can eg. stop unnecessary requests to the API.
     - we add a validation step before setting them
- where the query key is not supported
    - I think throwing an Error can be a bit excessive. I would've taken the approach of just omitting the key from the params to construct, and along with it emit a warning via `logger` to inform the consumer about the unsupported key.

Reply:
I would also simply warn if a key is unknown, it's not really an error.

Reply:
@emmenko I think we should throw an error here, because:

The users task is to pass configuration to `RequestBuilder`. He can either do that imperatively, or declaratively (using `parse`).

When doing it imperatively, the library will throw on unknown setters (e.g `service.perPag(4)` will throw as the setter is called `perPage`, not `perPag`). When using the declarative way the same rules should apply: `service.parse({ perPag: 4 }).build()`  should throw as well. I don't see how it would be helpful to allow passing unknown keys. It is very likely to be a bug if that happens. And it should be caught early.

Reply:
@adnasa 

> where a query is invalid => throw an error

This already happens naturally, as the `parse` function uses the setters behind the scenes. These check the passed values.

> where the query key is not supported

See the answer above.

Reply:
Ah I see what you mean. Makes sense yes 👍